In [5]:
# =========================
# ЯЧЕЙКА 0 — ТЕХНИЧЕСКАЯ: подключение к БД + read_df (обязательная)
# =========================
import sys
from pathlib import Path
import pandas as pd
from sqlalchemy import text

PROJECT_ROOT = Path().resolve()

# если ноутбук лежит в notebooks/, а src в корне — подними на уровень выше:
if not (PROJECT_ROOT / "src").exists() and (PROJECT_ROOT.parent / "src").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent

print("📌 PROJECT_ROOT =", PROJECT_ROOT)

if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from src.common.db import DBConfig, get_connection, set_search_path, test_connection

test_connection(verbose=True)
cfg = DBConfig.from_env()
print(f"✅ schema = {cfg.schema}")

def read_df(sql: str, params: dict | None = None) -> pd.DataFrame:
    with get_connection() as conn:
        set_search_path(conn)
        return pd.read_sql(text(sql), conn, params=params)

📌 PROJECT_ROOT = /Users/iriska/Downloads/mlbox_2
OK: connected to PostgreSQL | host=localhost port=55432 db=gpx_db schema=public
✅ schema = public


In [7]:
# =========================
# МОДУЛЬ C — ЯЧЕЙКА 1: сбор датасета из БД (segment_labels -> X, y)
# =========================

import json
import numpy as np
import pandas as pd

df = read_df("""
SELECT
  segment_id, track_id, segment_index, cluster_id,
  risk_label, evac_label, features
FROM segment_labels
WHERE risk_label IS NOT NULL AND evac_label IS NOT NULL
ORDER BY segment_id;
""")

print("✅ segment_labels loaded:", df.shape)
display(df.head(3))

def to_dict(x):
    if x is None:
        return {}
    if isinstance(x, dict):
        return x
    if isinstance(x, str):
        try:
            return json.loads(x)
        except Exception:
            return {}
    return {}

df["features_dict"] = df["features"].apply(to_dict)

def getv(d, key, default=np.nan):
    try:
        return d.get(key, default)
    except Exception:
        return default

def ctx_get(d, k, default=np.nan):
    ctx = d.get("ctx", {}) if isinstance(d.get("ctx", {}), dict) else {}
    return ctx.get(k, default)

# базовые фичи сегмента
df["distance_m"]    = df["features_dict"].apply(lambda d: getv(d, "distance_m"))
df["duration_s"]    = df["features_dict"].apply(lambda d: getv(d, "duration_s"))
df["avg_speed_mps"] = df["features_dict"].apply(lambda d: getv(d, "avg_speed_mps"))
df["elev_gain_m"]   = df["features_dict"].apply(lambda d: getv(d, "elev_gain_m"))
df["elev_loss_m"]   = df["features_dict"].apply(lambda d: getv(d, "elev_loss_m"))
df["elev_min_m"]    = df["features_dict"].apply(lambda d: getv(d, "elev_min_m"))
df["elev_max_m"]    = df["features_dict"].apply(lambda d: getv(d, "elev_max_m"))
df["slope"]         = df["features_dict"].apply(lambda d: getv(d, "slope"))
df["points"]        = df["features_dict"].apply(lambda d: getv(d, "points"))

# ctx фичи
df["temp_c"]                = df["features_dict"].apply(lambda d: ctx_get(d, "temp_c"))
df["nearby_waterway_total"] = df["features_dict"].apply(lambda d: ctx_get(d, "nearby_waterway_total", 0.0))
df["nearby_highway_total"]  = df["features_dict"].apply(lambda d: ctx_get(d, "nearby_highway_total", 0.0))
df["nearby_natural_total"]  = df["features_dict"].apply(lambda d: ctx_get(d, "nearby_natural_total", 0.0))
df["nearby_landuse_total"]  = df["features_dict"].apply(lambda d: ctx_get(d, "nearby_landuse_total", 0.0))
df["nearby_building_total"] = df["features_dict"].apply(lambda d: ctx_get(d, "nearby_building_total", 0.0))

# cluster_id как число (если NULL -> NaN)
df["cluster_id"] = pd.to_numeric(df["cluster_id"], errors="coerce")

feature_cols = [
    "distance_m","duration_s","avg_speed_mps",
    "elev_gain_m","elev_loss_m","elev_min_m","elev_max_m",
    "slope","points",
    "temp_c",
    "nearby_waterway_total","nearby_highway_total",
    "nearby_natural_total","nearby_landuse_total","nearby_building_total",
    "cluster_id"
]

X = df[feature_cols].apply(pd.to_numeric, errors="coerce").copy()
y_risk = df["risk_label"].astype(str)
y_evac = df["evac_label"].astype(str)

print("\n📌 X shape:", X.shape)
print("📌 y_risk:", y_risk.value_counts(dropna=False).to_dict())
print("📌 y_evac:", y_evac.value_counts(dropna=False).to_dict())

print("\n📊 NaN по признакам (это ок — дальше будет Imputer):")
display(X.isna().sum().to_frame("nan_cnt"))

✅ segment_labels loaded: (8663, 7)


segment_id  track_id  segment_index  cluster_id risk_label evac_label  \
0           1         2              0           4       fire       hard   
1           2         2              1           4       fire       easy   
2           3         2              2           2       fire     medium   

                                            features  
0  {'ctx': {'ctx_id': 1, 'temp_c': -6.2, 'matched...  
1  {'ctx': {'ctx_id': 1, 'temp_c': -6.2, 'matched...  
2  {'ctx': {'ctx_id': 1, 'temp_c': -6.2, 'matched...


📌 X shape: (8663, 16)
📌 y_risk: {'none': 8427, 'flood': 225, 'fire': 11}
📌 y_evac: {'easy': 5133, 'medium': 3267, 'hard': 263}

📊 NaN по признакам (это ок — дальше будет Imputer):


nan_cnt
distance_m                   0
duration_s                7469
avg_speed_mps             7503
elev_gain_m               2714
elev_loss_m               2714
elev_min_m                2712
elev_max_m                2712
slope                     2767
points                       0
temp_c                    8190
nearby_waterway_total        0
nearby_highway_total         0
nearby_natural_total         0
nearby_landuse_total         0
nearby_building_total        0
cluster_id                   0

In [8]:
# =========================
# МОДУЛЬ C — ЯЧЕЙКА 2: обучение моделей (risk_label)
# =========================

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
import joblib
from pathlib import Path

# --- train / test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y_risk,
    test_size=0.25,
    random_state=42,
    stratify=y_risk
)

# --- общий препроцессинг
prep = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

Xtr = prep.fit_transform(X_train)
Xte = prep.transform(X_test)

models = {
    "logreg": LogisticRegression(
        max_iter=1000,
        class_weight="balanced"
    ),
    "rf": RandomForestClassifier(
        n_estimators=200,
        random_state=42,
        class_weight="balanced"
    ),
    "gboost": GradientBoostingClassifier(
        random_state=42
    )
}

results = []

for name, model in models.items():
    model.fit(Xtr, y_train)
    preds = model.predict(Xte)

    f1 = f1_score(y_test, preds, average="macro")
    print(f"\n📌 {name.upper()} — F1_macro = {f1:.4f}")
    print(classification_report(y_test, preds))

    results.append({
        "model": name,
        "f1_macro": f1
    })

# --- сохраняем лучший
best = max(results, key=lambda x: x["f1_macro"])
best_name = best["model"]
best_model = models[best_name]

print(f"\n🏆 Лучшая модель: {best_name}")

Path("models").mkdir(exist_ok=True)

joblib.dump(best_model, f"models/risk_model_{best_name}.joblib")
joblib.dump(prep, "models/preprocess.joblib")

print("💾 Модель и препроцессинг сохранены в папку models/")


📌 LOGREG — F1_macro = 0.9304
              precision    recall  f1-score   support

        fire       1.00      0.67      0.80         3
       flood       0.98      1.00      0.99        56
        none       1.00      1.00      1.00      2107

    accuracy                           1.00      2166
   macro avg       0.99      0.89      0.93      2166
weighted avg       1.00      1.00      1.00      2166


📌 RF — F1_macro = 0.9304
              precision    recall  f1-score   support

        fire       1.00      0.67      0.80         3
       flood       0.98      1.00      0.99        56
        none       1.00      1.00      1.00      2107

    accuracy                           1.00      2166
   macro avg       0.99      0.89      0.93      2166
weighted avg       1.00      1.00      1.00      2166


📌 GBOOST — F1_macro = 0.9494
              precision    recall  f1-score   support

        fire       0.75      1.00      0.86         3
       flood       1.00      0.98      0.99

📘 МОДУЛЬ C — Обучение моделей классификации и автоматизация (ML Agent)

Назначение модуля

Цель модуля C — построение автоматизированного ML-агента, который на основе размеченных данных из базы данных:
	•	формирует обучающий датасет,
	•	обучает несколько моделей классификации,
	•	оценивает качество моделей,
	•	выбирает лучшую модель,
	•	сохраняет обученные артефакты для дальнейшего использования в API и приложении.

Модуль использует результаты Модуля B (разметка данных) и не требует создания дополнительных таблиц в базе данных.

⸻

🧩 ЯЧЕЙКА 1 — Сбор обучающего датасета из БД

Что делаем

На данном этапе производится загрузка размеченных сегментов маршрутов из таблицы segment_labels.

Используются только те записи, для которых:
	•	задана метка риска (risk_label),
	•	задана сложность эвакуации (evac_label).

Это гарантирует корректность и полноту обучающих данных.

Почему так
	•	База данных является единственным источником истины.
	•	Агент не хранит временные данные — датасет формируется при каждом запуске.
	•	Такой подход имитирует работу production-агента.

Результат

Получен датафрейм из 8663 сегментов, содержащий:
	•	идентификаторы сегментов,
	•	кластер,
	•	метки риска и эвакуации,
	•	JSON-признаки сегмента.

⸻

🧩 ЯЧЕЙКА 2 — Унификация и безопасный парсинг признаков

Что делаем

JSON-поле features безопасно преобразуется в Python-словарь.

Решается важная кроссплатформенная проблема:
	•	в Linux/macOS JSONB приходит как dict,
	•	в Windows — как str.

Используется универсальная функция to_dict, корректно обрабатывающая оба случая.

Почему так
	•	Код становится переносимым между ОС.
	•	Исключаются ошибки при обучении моделей и деплое.

Результат

Для каждого сегмента гарантированно доступна структура features_dict.

⸻

🧩 ЯЧЕЙКА 3 — Формирование признакового пространства (X)

Что делаем

Из features_dict извлекаются:

Базовые признаки сегмента:
	•	расстояние,
	•	длительность,
	•	средняя скорость,
	•	перепад высот,
	•	уклон,
	•	количество точек.

Контекстные признаки:
	•	температура воздуха,
	•	плотность дорог,
	•	наличие водоёмов,
	•	природные и застроенные зоны.

Также добавляется cluster_id, полученный на этапе кластеризации.

Почему так
	•	Признаки отражают физическую сложность участка и географический контекст.
	•	Модель обучается на тех же данных, которые использовались при разметке → логическая согласованность.

Результат

Сформирована матрица признаков X размерности:

(8663, 16)

🧩 ЯЧЕЙКА 4 — Формирование целевых переменных (y)

Что делаем

Формируются две независимые задачи классификации:
	1.	y_risk — тип риска:
	•	none
	•	flood
	•	fire
	2.	y_evac — сложность эвакуации:
	•	easy
	•	medium
	•	hard

Почему так
	•	Это две разные задачи, требующие отдельных моделей.
	•	Такой подход упрощает интерпретацию и масштабирование.

Результат

Подтверждено корректное распределение классов и отсутствие пропусков.

⸻

🧩 ЯЧЕЙКА 5 — Анализ пропусков и подготовка к обучению

Что делаем

Анализируется количество NaN по каждому признаку.

Почему это нормально
	•	Отсутствие контекста (погоды, окружения) — реальный сценарий.
	•	В дальнейшем используется Imputer, поэтому данные не теряются.

Результат

Данные признаны пригодными для обучения без дополнительной очистки.

⸻

🧩 ЯЧЕЙКА 6 — Обучение моделей классификации

Что делаем

Обучаются минимум три алгоритма классификации, как требуется по заданию:
	•	Logistic Regression
	•	Random Forest
	•	Gradient Boosting

Используется единый preprocessing-pipeline:
	•	заполнение пропусков,
	•	масштабирование признаков.

Почему так
	•	Сравнение разных подходов повышает устойчивость решения.
	•	Исключается зависимость от одной модели.

⸻

🧩 ЯЧЕЙКА 7 — Оценка качества моделей

Метрика

Используется F1_macro, так как:
	•	классы несбалансированы,
	•	важно одинаково учитывать редкие риски.

Результат

Все модели показали высокий результат:

F1_macro ≈ 0.93

🧩 ЯЧЕЙКА 8 — Выбор и сохранение лучшей модели

Что делаем
	•	выбирается лучшая модель по метрике,
	•	сохраняются:
	•	модель,
	•	preprocessing-pipeline.

Почему так
	•	Модель готова к использованию в API,
	•	агент становится воспроизводимым,
	•	обеспечивается непрерывное обучение.

Результат

Артефакты сохранены в папку models/.

⸻

✅ Итог по модулю C

В рамках модуля C реализован полноценный ML-агент, который:
	•	автоматически формирует датасет из БД,
	•	обучает несколько моделей классификации,
	•	оценивает их качество,
	•	выбирает лучшую модель,
	•	сохраняет результаты для дальнейшего использования.

Модуль полностью соответствует требованиям задания и готов к интеграции с API и пользовательским приложением.